### PR TITLE
Add implementation of training based on pseudo-likelihood

### PR DESCRIPTION
**Pull request by pamatteo**

### CODE REVIEW

File: src/base/maputil.h [hunk 1]
```diff
@@ -44,13 +44,13 @@ struct IntPair {
   int second;
 };
 
-namespace std { namespace tr1 {
+/*namespace std { namespace tr1 {
```
Comment (vraychev) [inline src/base/maputil.h:47]: Do not comment out code. Either delete or keep.

File: src/inference/graph_inference.cpp [hunk 2]
```diff
@@ -17,11 +17,13 @@
 #include "graph_inference.h"
 
 #include <stdio.h>
+#include <math.h>
 #include <queue>
 #include <unordered_set>
 
 #include "stringprintf.h"
 #include "glog/logging.h"
+#include "gflags/gflags.h"
```
Comment (vraychev) [inline src/inference/graph_inference.cpp:26]: Seems unneeded

File: src/inference/graph_inference.cpp [hunk 3]
```diff
@@ -341,6 +349,28 @@ class GraphNodeAssignment : public Nice2Assignment {
     return label_set_->GetLabelName(label_id);
   }
 
+  // Returns conditional probability of a node
```
Comment (vraychev) [inline src/inference/graph_inference.cpp:352]: It is not of a node, it is of a label | conditioned on what?

File: src/base/maputil.h (hunk 1 above)
Comment (pamatteo) [inline src/base/maputil.h:47]: Ok. Got it.

File: src/inference/graph_inference.cpp [hunk 4]
```diff
@@ -348,19 +378,24 @@ class GraphNodeAssignment : public Nice2Assignment {
...
-          assignments_[arc.node_a].label,
-          assignments_[arc.node_b].label,
-          arc.type);
-      auto feature_it = features.find(feature);
-      if (feature_it != features.end()) {
-        sum += feature_it->second.getValue();
-      }
-    }
-    return sum;
+        const GraphInference::FeaturesMap& features = fweights.features_;
```
Comment (vraychev) [inline src/inference/graph_inference.cpp:387]: Did you just replace spaces with tabs?

File: src/inference/graph_inference.cpp [hunk 5]
```diff
@@ -537,6 +572,32 @@ class GraphNodeAssignment : public Nice2Assignment {
...
+          arc.type);
+      (*affected_features)[feature] += gradient_weight;
+    }
+  }
+
+  void RegularizeAffectedFeature(
+      GraphInference::SimpleFeaturesMap* affected_features,
+      double regularizer) const {
+    for (auto feature : (*affected_features)) {
+      feature.second += (feature.second / regularizer);
```
Comment (vraychev) [inline src/inference/graph_inference.cpp:597]: Did you means = instead of += ?

File: src/inference/graph_inference.cpp [hunk 6]
```diff
@@ -537,6 +572,32 @@ class GraphNodeAssignment : public Nice2Assignment {
...
+          assignments[arc.node_b].label,
+          arc.type);
+      (*affected_features)[feature] += gradient_weight;
+    }
+  }
+
+  void RegularizeAffectedFeature(
+      GraphInference::SimpleFeaturesMap* affected_features,
+      double regularizer) const {
+    for (auto feature : (*affected_features)) {
```
Comment (vraychev) [inline src/inference/graph_inference.cpp:596]: You know this is making a copy. You need auto & 

Alternatively, write the loop yourself to avoid confusion.

File: src/inference/graph_inference.cpp [hunk 7]
```diff
@@ -1049,29 +1104,16 @@ double GraphInference::GetAssignmentScore(const Nice2Assignment* assignment) con
...
-    const Nice2Assignment* assignment,
-    double learning_rate,
-    PrecisionStats* stats) {
-  const GraphNodeAssignment* a = static_cast<const GraphNodeAssignment*>(assignment);
-
-  GraphNodeAssignment new_assignment(*a);
-  new_assignment.SetUpEqualityPenalty(svm_margin_);
-  PerformAssignmentOptimization(&new_assignment);
+void GraphInference::UpdateStats(
+    const GraphNodeAssignment assignment,
```
Comment (vraychev) [inline src/inference/graph_inference.cpp:1108]: With this you are copying objects. Did you want const GraphNodeAssignment&  ?

File: src/inference/graph_inference.cpp [hunk 8]
```diff
@@ -714,12 +775,6 @@ class GraphNodeAssignment : public Nice2Assignment {
   }
 
 private:
-  struct Assignment {
```
Comment (vraychev) [inline src/inference/graph_inference.cpp:717]: Not sure why you made this public.

File: src/inference/inference.h [hunk 9]
```diff
@@ -20,7 +20,7 @@
 #include <mutex>
 #include <string>
 #include <map>
-#include "jsoncpp/json/json.h"
+#include "json/json.h"
```
Comment (vraychev) [inline src/inference/inference.h:23]: Do you break the linux build?

File: src/inference/graph_inference.cpp (hunk 2 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp:26]: On my machine it wasn't compiling without that include. I will double check to be sure that it is necessary.

File: src/inference/graph_inference.cpp [hunk 10]
```diff
@@ -1101,10 +1166,63 @@ void GraphInference::SSVMLearn(
...
+      double normalization_constant = a->GetNodeNormalizationConstantGivenNeighbours(*this, i, &candidates);
+      for (int label : candidates) {
+        double marginal_probability = a->GetNodeConditionalProbabilityGivenNeighbours(*this, i, normalization_constant, label);
+        a->GetNeghbouringAffectedFeature(&affected_features, i, label, -learning_rate * marginal_probability);
       }
     }
   }
+
+  a->GetAffectedFeatures(&affected_features, beam_size_ * learning_rate);
+  a->RegularizeAffectedFeature(&affected_features, num_training_samples * pow(pl_regularizer_, 2));
```
Comment (vraychev) [inline src/inference/graph_inference.cpp:1217]: This needs explanation. How is num_training_samples needed and what is this power? Did you want pow on 0.5 or 2 ?

File: src/inference/graph_inference.cpp (hunk 3 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp:352]: You are right, and it is conditioned on its neighbours.

File: src/inference/graph_inference.cpp (hunk 4 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp:387]: Nope, this gets the node score based on the assignment passed as argument. I use it when I compute the normalisation constant and I have to change the label of a certain node.

Comment (pamatteo) [inline src/inference/graph_inference.cpp:387]: But you are right it detects also the change of the spaces into tabs.

File: src/inference/graph_inference.cpp (hunk 5 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp:597]: Nope I meant += the same as in the GetAffectedFeatures

File: src/inference/graph_inference.cpp (hunk 6 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp:596]: Ok I will add the ampersand

File: src/inference/graph_inference.cpp (hunk 7 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp:1108]: Yeah, you are right it's better passing it via address.

File: src/inference/graph_inference.cpp (hunk 8 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp:717]: Because I needed to access the field inside the assignment struct, in order to change the label of a node.

File: src/inference/inference.h (hunk 9 above)
Comment (pamatteo) [inline src/inference/inference.h:23]: This would break the linux build, I will add a flag. Thanks for noticing it.

File: src/inference/graph_inference.cpp (hunk 10 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp:1217]: The num_training_samples is need to normalise the gradient of a feature. And I wanted pow on 2.
